File: Haskell/Template Haskell/LetRec.hs
module LetRec where

import Control.Monad
import Language.Haskell.TH


-- |@letrec n@ defines a \"let rec\" function for @n@ recursive bindings.  In
-- particular, it takes @n+1@ arguments: the definitions of each of the @n@
-- recursive bindings as functions of all @n@ values, and the \"in\" part of
-- the let expression, as a function of the @n@ recursive bindings.
--
-- For example, @letrec 3@ generates the following code (with uglier
-- variable names):
-- 
-- > \f g h expr -> let x = f x y z
-- >                    y = g x y z
-- >                    z = h x y z
-- >                 in expr x y z
letrec :: Int -> DecQ
letrec n = do
    fs <- replicateM n (newName "f")
    xs <- replicateM n (newName "x")
    expr <- newName "expr"
    
    let xDecs = 
            [ valD (varP x) (normalB (appsE (varE f : map varE xs))) []
            | (f,x) <- zip fs xs
            ]
    
    lamE (map varP fs ++ [varP expr]) (letE xDecs (appsE (varE expr : map varE xs)))
    

File: Scala/letRec.scala
def letRec[X](x: (=>X)=>X): X =
{
    lazy val xx:X = x(xx)
    xx
}

def letRec[X,Y](x: (=>X,=>Y)=>X,
                y: (=>X,=>Y)=>Y): (X,Y) =
{
    object rec {
        lazy val xx:X = x(xx,yy)
        lazy val yy:Y = y(xx,yy)
    }
    (rec.xx,rec.yy)
}

def letRec[X,Y,Z](x: (=>X,=>Y,=>Z)=>X,
                  y: (=>X,=>Y,=>Z)=>Y,
                  z: (=>X,=>Y,=>Z)=>Z): (X,Y,Z) =
{
    object rec {
        lazy val xx:X = x(xx,yy,zz)
        lazy val yy:Y = y(xx,yy,zz)
        lazy val zz:Z = z(xx,yy,zz)
    }
    (rec.xx,rec.yy,rec.zz)
}

def letRec[W,X,Y,Z](w: (=>W,=>X,=>Y,=>Z)=>W,
                    x: (=>W,=>X,=>Y,=>Z)=>X,
                    y: (=>W,=>X,=>Y,=>Z)=>Y,
                    z: (=>W,=>X,=>Y,=>Z)=>Z): (W,X,Y,Z) =
{
    object rec {
        lazy val ww:W = w(ww,xx,yy,zz)
        lazy val xx:X = x(ww,xx,yy,zz)
        lazy val yy:Y = y(ww,xx,yy,zz)
        lazy val zz:Z = z(ww,xx,yy,zz)
    }
    (rec.ww,rec.xx,rec.yy,rec.zz)
}

